### PR TITLE
*: deprecate DateArithInterval

### PR DESCRIPTION
**Pull request by zyguan**
`DateArithInterval` will prevent its inner interval expression from being evaluated, which leads to incorrect results for function calls like `adddate('2011-11-11', least(1,2))`.

Here is the original logic:
```go
func builtinDateArith(args []types.Datum, ctx context.Context) (d types.Datum, err error) {
	// Op is used for distinguishing date_add and date_sub.
	// args[0] -> Op
	// args[1] -> Date
	// args[2] -> DateArithInterval
	// health check for date and interval
	if args[1].IsNull() {
		return d, nil
	}
	nodeDate := args[1]
	nodeInterval := args[2].GetInterface().(ast.DateArithInterval)
	nodeIntervalIntervalDatum := nodeInterval.Interval.GetDatum()
	if nodeIntervalIntervalDatum.IsNull() {
		return d, nil
	}
...
```

`nodeInterval.Interval` may be a `FuncCallExpr` that not be evaluated by `FoldConstant` (because `DateArithInterval` is not a `ScalarFunction`).

Thus, `nodeIntervalIntervalDatum.IsNull()` returns `true` for aforementioned function call.

This PR can solve the issue by disabling the `DateArithInterval`. Is there any better solution?

### CODE REVIEW

**Comment by shenli** [comment]
Thanks! @zyguan 
@tiancaiamao @hanfei1991 PTAL

File: expression/builtin_time_test.go [hunk 1]
```diff
@@ -533,3 +534,35 @@ func (s *testEvaluatorSuite) TestYearWeek(c *C) {
...
+	date := []string{"2016-12-31", "2017-01-01"}
+
+	var (
+		args []types.Datum
+		v    types.Datum
+		err  error
+	)
+
+	args = types.MakeDatums(ast.DateAdd, date[0], 1, "DAY")
+	v, err = builtinDateArith(args, s.ctx)
```
Comment (zimulala) [inline expression/builtin_time_test.go]: Please using `:=` instead of `=`, then remove line543-line548.

File: parser/lexer_test.go [hunk 2]
```diff
@@ -213,13 +213,13 @@ func (s *testLexerSuite) TestIdentifier(c *C) {
 func (s *testLexerSuite) TestSpecialComment(c *C) {
 	l := NewScanner("/*!40101 select\n5*/")
 	tok, pos, lit := l.scan()
-	fmt.Println(tok, pos, lit)
+	//fmt.Println(tok, pos, lit)
```
Comment (zimulala) [inline parser/lexer_test.go]: It is better to remove it.

File: parser/parser_test.go [hunk 3]
```diff
@@ -769,9 +775,9 @@ func (s *testParserSuite) TestBuiltin(c *C) {
 		{`select subdate("2011-11-11 10:10:10.123456", interval "11 10:10" day_minute)`, true},
 		{`select subdate("2011-11-11 10:10:10.123456", interval "11 10" day_hour)`, true},
 		{`select subdate("2011-11-11 10:10:10.123456", interval "11-11" year_month)`, true},
-		{`select adddate("2011-11-11 10:10:10.123456", 10)`, true},
-		{`select adddate("2011-11-11 10:10:10.123456", 0.10)`, true},
-		{`select adddate("2011-11-11 10:10:10.123456", "11,11")`, true},
+		{`select subdate("2011-11-11 10:10:10.123456", 10)`, true},
```
Comment (hanfei1991) [inline parser/parser_test.go]: why change add to sub?

File: expression/builtin_time.go [hunk 4]
```diff
@@ -638,15 +638,16 @@ func builtinDateArith(args []types.Datum, ctx context.Context) (d types.Datum, e
 	// Op is used for distinguishing date_add and date_sub.
 	// args[0] -> Op
```
Comment (hanfei1991) [inline expression/builtin_time.go]: If you like, you can remove the args[0], and use a builtin function factory instead. Like what the `compareFuncFactory` do, then we will not rely on any ugly things any more :)

File: parser/lexer_test.go (hunk 2 above)
Comment (zimulala) [inline parser/lexer_test.go]: It is better to remove it.

File: expression/builtin_time_test.go (hunk 1 above)
Comment (zimulala) [inline expression/builtin_time_test.go]: Please using `:=` instead of `=`, then remove line543-line548.

File: parser/parser_test.go (hunk 3 above)
Comment (zyguan) [inline parser/parser_test.go]: I think it's a typo here. `adddate` has been tested [here](https://github.com/pingcap/tidb/pull/2368/files/e38dccc5293f3180052caee012983cdceaa96ed0#diff-27c45ca411f005e1b8796b12fb53e26cL725).

File: expression/builtin_time.go (hunk 4 above)
Comment (zyguan) [inline expression/builtin_time.go]: Thanks @hanfei1991 , I'll give it a try.

File: expression/builtin_time_test.go (hunk 1 above)
Comment (zyguan) [inline expression/builtin_time_test.go]: copy :)

File: expression/builtin_time.go (hunk 4 above)
Comment (zyguan) [inline expression/builtin_time.go]: Hi @hanfei1991 , I think such a refactoring is not strongly related to the topic of this PR. Can I do it in the next PR?

**Comment by tiancaiamao** [comment]
PTAL @hanfei1991 @zimulala

**Comment by hanfei1991** [comment]
LGTM

**Pull request merged**